From 6b7f6d39fe68ab3c55880f51a6769ee73fdc8295 Mon Sep 17 00:00:00 2001
From: Tushar Behera <tushar.behera@linaro.org>
Date: Wed, 15 May 2013 09:58:15 +0530
Subject: [PATCH] ARM: EXYNOS: Set 64-bit DMA mask for EXYNOS5250
Patch-Mainline: Not yet

ARM: EXYNOS: Set 64-bit DMA mask for EXYNOS5250

This is required when LPAE is enabled.

Signed-off-by: Tushar Behera <tushar.behera@linaro.org>
Signed-off-by: Alexander Graf <agraf@suse.de>
---
 arch/arm/mach-exynos/mach-exynos5-dt.c |   24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

Index: linux-3.12-rc2-master/arch/arm/mach-exynos/mach-exynos5-dt.c
===================================================================
--- linux-3.12-rc2-master.orig/arch/arm/mach-exynos/mach-exynos5-dt.c
+++ linux-3.12-rc2-master/arch/arm/mach-exynos/mach-exynos5-dt.c
@@ -14,6 +14,7 @@
 #include <linux/memblock.h>
 #include <linux/io.h>
 #include <linux/clocksource.h>
+#include <linux/dma-mapping.h>
 
 #include <asm/mach/arch.h>
 #include <mach/regs-pmu.h>
@@ -23,6 +24,26 @@
 
 #include "common.h"
 
+static u64 dma_mask64 = DMA_BIT_MASK(64);
+
+static int exynos5250_platform_notifier(struct notifier_block *nb,
+				  unsigned long event, void *__dev)
+{
+	struct device *dev = __dev;
+
+	if (event != BUS_NOTIFY_ADD_DEVICE)
+		return NOTIFY_DONE;
+
+	dev->dma_mask = &dma_mask64;
+	dev->coherent_dma_mask = DMA_BIT_MASK(64);
+
+	return NOTIFY_OK;
+}
+
+static struct notifier_block exynos5250_platform_nb = {
+	.notifier_call = exynos5250_platform_notifier,
+};
+
 static void __init exynos5_dt_machine_init(void)
 {
 	struct device_node *i2c_np;
@@ -47,6 +68,9 @@ static void __init exynos5_dt_machine_in
 		}
 	}
 
+	if (of_machine_is_compatible("samsung,exynos5250"))
+		bus_register_notifier(&platform_bus_type, &exynos5250_platform_nb);
+
 	of_platform_populate(NULL, of_default_bus_match_table, NULL, NULL);
 }
 
